### PR TITLE
Python Lab: Make validation work on template backed levels

### PR DESCRIPTION
**Pull request by molly-moen**
Note: about 100 lines of changes in this PR are level files, feel free to ignore those.

We discovered a flaw in the way we were handling validation late last week in relation to template backed levels. Previously, validation was just part of start sources, which meant for template-backed levels you could only have validation on the template level, not the level itself. However, we want to be able to define validation on the actual level, so we can test different things on different levels (this is how Java Lab works). In Java Lab, we store the validation separately from the start sources to enable this (Java Lab also works differently with validation in that it is only accessed on the backend, as we send it over to Javabuilder).

After thinking through options, I decided to go with a similar approach here. There are 2 flows to think about: writing start code and using the level as a student.

The start code flow is:
- The editor still behaves the same way; levelbuilders can add a single validation file to the project anywhere they want. 
- When we save the start sources, we take the validation file (if it exists), pull it out of start sources, and save it as a separate field on the level. We also mark it as not open or active to make the start sources clean.
- When levelbuilders load a level with a validation file, we add it into the sources again so they can edit and view it. It will always start as not open unless we hear that is annoying to levelbuilders.
- Levelbuilders must add a validation file, save, and reload to see the validate button (this is consistent with current behavior, but is tracked as a follow-up item). Once they've done so, the validate button will always run the most recent validation code in the editor, not the validation from the level.

The user flow is:
- Start code is handled identically, there is just now no validation file to ignore.
- The validation button will show up if there is validation on the level, same as before. The condition is easier to determine now because we have a field on the level (see `pythonlab.rb`)
- When running validation code, we pass in the validation file from the level as a new parameter. That file gets added to the source by `pyodideRunner`. Since we send back the files in the virtual pyodide system (to be able to handle file writing), I added a new condition to skip certain filenames, and we skip sending back the validation file.
- I also updated the duplicate filename checker to not allow using the validation file name. We don't allow it in any folder to make it easier to skip saving it on the pyodide side. We could possibly clean this up if we decide to restrict the validation file to just the root folder.
- I also updated the next file id check to also include the id of the validation file.

This work means that now the user also does not need to start over to get updated validation for the level, they just need to refresh the page.

## Links

- jira ticket: [CT-827](https://codedotorg.atlassian.net/browse/CT-827)
- [slack thread](https://codedotorg.slack.com/archives/C06JELCAPT9/p1727461892931889)

## Testing story
Tested locally that we can have levels both with and without validation, that template-backed levels now use the validation from the level, and start mode works as expected. I wanted to get this done quickly to unblock curriculum, hence the multiple follow-up tickets.

## Follow-up work
There's some clean up and improvements we can do here.
- [CT-830](https://codedotorg.atlassian.net/browse/CT-830): improve start mode flow to not require a refresh
- [CT-831](https://codedotorg.atlassian.net/browse/CT-831): once this is on levelbuilder, this is the repair step to get existing levels to use the new format. We just need to click save on start mode for each level. There are around 30 of these levels now, so it's a bit of work but it will be worth it for long-term stability of this feature.
- [CT-832](https://codedotorg.atlassian.net/browse/CT-832): once we've done the clean up, we may be able to get rid of some outdated logic re: validation files
- [CT-833](https://codedotorg.atlassian.net/browse/CT-833): Consider restricting validation files to the root of the project.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/codebridge/FileBrowser/hooks/useHandleFileUpload.ts [hunk 1]
```diff
@@ -20,6 +22,10 @@ type handleFileUploadArgs = {
 
 export const useHandleFileUpload = (files: Record<string, ProjectFile>) => {
   const appName = useAppSelector(state => state.lab.levelProperties?.appName);
+  const isStartMode = getAppOptionsEditBlocks() === START_SOURCES;
+  const validationFile = useAppSelector(
+    state => state.lab.levelProperties?.validationFile
+  );
```
Comment (Erin007) [inline apps/src/codebridge/FileBrowser/hooks/useHandleFileUpload.ts:28]: just double checking - even if there is more than one test, there will always only be one validation file?

File: apps/src/codebridge/hooks/useSource.ts [hunk 2]
```diff
@@ -65,24 +72,35 @@ export const useSource = (defaultSources: ProjectSources) => {
...
     levelStartSource,
+    validationFile,
   ]);
 
   useEffect(() => {
     if (isStartMode) {
       header.showLevelBuilderSaveButton(() => {
-        return {start_sources: source};
+        const {parsedSource, validationFile} =
+          prepareSourceForLevelbuilderSave(source);
```
Comment (Erin007) [inline apps/src/codebridge/hooks/useSource.ts:102]: nice!

File: apps/src/codebridge/utils/fileUtils.ts [hunk 3]
```diff
@@ -35,3 +39,61 @@ export function getFileIconNameAndStyle(file: ProjectFile): {
...
+    )
+  );
+  let validationFile =
+    Object.values(source.files).find(
+      f => f.type === ProjectFileType.VALIDATION
+    ) || null;
+  let openFiles = source.openFiles;
+  if (validationFile && source.openFiles?.includes(validationFile.id)) {
+    openFiles = source.openFiles.filter(id => id !== validationFile?.id);
+    //validationFile.open = false;
```
Comment (Erin007) [inline apps/src/codebridge/utils/fileUtils.ts]: do you need this?

File: apps/src/pythonlab/pyodideRunner.ts [hunk 4]
```diff
@@ -56,19 +65,25 @@ export function stopPythonCode() {
...
+  validationFile?: ProjectFile
 ) {
-  // If the project has a validation file, we just run those tests.
-  const validationFile = Object.values(source.files).find(
-    f => f.type === ProjectFileType.VALIDATION
-  );
-  if (validationFile) {
-    // We only support one validation file. If somehow there is more than one, just run the first one.
+  // We default to using the validation file passed in. If it does not exist,
+  // we check the source for the validation file (this is the case in start mode).
```
Comment (Erin007) [inline apps/src/pythonlab/pyodideRunner.ts:72]: ah. this answers my question above about multiple validation files...

File: dashboard/app/models/levels/pythonlab.rb [hunk 5]
```diff
@@ -81,8 +82,7 @@ def clean_up_predict_settings
   # Return the validation condition for this level. If the level has a validation file, the condition
   # is that all tests passed. If there is no validation file, there are no conditions.
   def get_validations
-    has_validation = start_sources && start_sources["files"]&.any? {|(_, file)| file["type"] == 'validation'}
-    if has_validation
+    if validation_file
```
Comment (Erin007) [inline dashboard/app/models/levels/pythonlab.rb:85]: yay! this reads cleaner than `has_validation`

**Review by Erin007 — APPROVED** [review]

File: apps/src/codebridge/utils/fileUtils.ts (hunk 3 above)
Comment (molly-moen) [reply apps/src/codebridge/utils/fileUtils.ts]: good catch, nope because of the next line

File: apps/src/pythonlab/pyodideRunner.ts (hunk 4 above)
Comment (molly-moen) [reply apps/src/pythonlab/pyodideRunner.ts:72]: yep there's only ever one

**Pull request merged**